### ISSUE TITLE
RepeatAttribute

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit#478]
NUnit 2.6 has a RepeatAttribute. It's currently #ifed out of our 3.0 source. We need to decide if we are going to implement it as is or make some changes. People ask for it (sometimes for bad reasons) so I don't think it should be simply dropped.

Current semantics are that the test is run n times or until it fails, whichever comes first. IOW, it has to pass all n times in order to pass.

We have had requests in the past for only repeating the test in the case of failure. IOW, it would be more like a Retry attribute.

Conceivably, we could generalize this to repeating the test up to n times subject to some rules specified in additional arguments. We could expose the options to the user if we liked or we could simply use such a generalized repeat facility as a base class for various attributes.

We also have a request for ThreadedRepeat, which would run the tests n times on different threads. See #5.

I'm marking this issue as status:design. We should first come up with a spec for what we are going to support and then incorporate the result into 3.0.

**Comment by rprouse** [comment]
The following are use cases from one developer on the mailing list for the repeat attribute. It sounds like the third is better suited to ThreadedRepeat although all of them might be better served with ThreadedRepeat #5

### Bug reproducing (multi-threaded) issue. 
To reproduce a bug, developer implements test. Sometimes (when the bug related to multi-threaded issue) we need execute the test many times to reproduce the bug. When bug reproduced, developer fixes found problem and execute the test again (with Repeat) to check implemented fix.

### Search "failed" input data for random failed test.
A test generates random input for tested code and sometimes fails. To found the input data we execute test with attribute Repeat. When the concrete input found, we add relevant test and delete "Repeat" from original test.

### Testing "multi-threaded" code.
To be sure if a concrete code works in multi-threaded (multi-process) mode, I implement test and permanently execute the test with Repeat(<enough big number>).

**Comment by CharliePoole** [comment]
I'll add to the test cases an additional use I had originally envisioned, which is similar to the "random" use case but a bit broader. I imagined that repeat would be used by tests that pull different data upon each execution, perhaps from an array, perhaps from a file. This was before we had parameterized tests, of course.

This could still be used in any case where parameterized tests won't work. In particular, it could be used to simulate "dynamic" data - i.e. data that's not available at load time.

**Comment by CharliePoole** [comment]
@rprouse 's use cases apply particularly to RepeatAttribute as defined. That is, an attribute that causes repetition unless the test fails.

Some other people are talking about Repeat as if it worked in the reverse way, that is, to keep repeating until the test passes! I call this RetryAttribute and I think that's the one to worry about a little. If your tests only pass after you retry them several times, then I'd say there's something wrong.

Finally, so-called ThreadedRepeat is an entirely different animal. There's no repetition at all, but you have several instances of the test running in parallel. We're not doing that for 3.0, but it's listed for a future release.

Note that both Repeat and Retry are actually _required_ to run all tests in sequence, so they will have to run on the same thread.

If we want to talk about Retry, it should be on a separate issue - in fact we may have one either here or on Launchpad.

Given all the use cases, I'd be inclined to put Repeat into the first beta if the underpinnings are still there as I think they are. If not, we could do it in the second beta.

**Comment by CharliePoole** [comment]
Retry has been added as issue #585

**Comment by CharliePoole** [comment]
Since Retry and ThreadedRepeat now have their own issues, there's no more design work to do on this one. It should be implemented to duplicate what we already have in NUnit 2.x. That is, the test, including Setup and TearDown is run the number of times specified or until it fails. Only one result is reported - either success or failure. It would be good to indicate in the message on which repetition it failed. Further enhancements can always be requested.
